src/NUnitEngine/nunit.engine/Services/DomainManager.cs

```diff
@@ -145,63 +145,62 @@ public void Unload(AppDomain domain)
         #region Nested DomainUnloader Class
         class DomainUnloader
         {
-            private Thread thread;
-            private AppDomain domain;
+            private Thread _unloadThread;
+            private AppDomain _domain;
+            private Exception _unloadException;
 
             public DomainUnloader(AppDomain domain)
             {
-                this.domain = domain;
+                _domain = domain;
             }
 
             public void Unload()
             {
-                string domainName;
-                try
-                {
-                    domainName = "UNKNOWN";//domain.FriendlyName;
-                }
-                catch (AppDomainUnloadedException)
-                {
-                    return;
-                }
+                _unloadThread = new Thread(new ThreadStart(UnloadOnThread));
 
-                log.Info("Unloading AppDomain " + domainName);
+                _unloadThread.Start();
 
-                thread = new Thread(new ThreadStart(UnloadOnThread));
-                thread.Start();
-                if (!thread.Join(30000))
+                if (!_unloadThread.Join(30000))
                 {
-                    log.Error("Unable to unload AppDomain {0}, Unload thread timed out", domainName);
-                    thread.Abort();
+                    string msg = "Unable to unload AppDomain, Unload thread timed out";
+
+                    log.Error(msg);
+                    _unloadThread.Abort();
+
+                    throw new NUnitEngineException(msg);
                 }
+
+                if (_unloadException != null)
+                    throw new NUnitEngineException("Exception encountered unloading AppDomain", _unloadException);
```

Comment:
This is a good change - our test suite started to fail with this in place. Unfortunately, it is really hard to get the _unloadException. I propose to change the exception to:
`throw new NUnitEngineException("Exception encountered unloading AppDomain: " + _unloadException.Message, _unloadException);` - or even add more of the stack trace.

I tried to get more details out of the test runner by using command line option `--trace=Info`, but the only trace of the log statements were found in `InternalTrace.15632.log`
```

17:48:04.787 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:17.747 Warning [ 1] ProcessRunner: Failed to unload the remote runner. Exception encountered unloading AppDomain
17:48:17.748 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:17.811 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:17.845 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:17.902 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:17.950 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:18.051 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:18.075 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:18.087 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:18.118 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:18.143 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:18.155 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:18.220 Info  [ 1] ProcessRunner: Unloading remote runner
17:48:18.226 Info  [ 1] ProcessRunner: Unloading remote runner
```
I execute tests in multiple test assemblies, so as you can see, I had multiple test runners. I could not find anything to determine which of my test assemblies caused the problem - here also a little more info would be nice...